generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt

```diff
@@ -53,13 +58,31 @@ class Chat(private val model: GenerativeModel, val history: MutableList<Content>
    * @throws InvalidStateException if the prompt is not coming from the 'user' role
    * @throws InvalidStateException if the [Chat] instance has an active request.
    */
-  suspend fun sendMessage(prompt: Content): GenerateContentResponse {
-    prompt.assertComesFromUser()
+  suspend fun sendMessage(inputPrompt: Content): GenerateContentResponse {
+    inputPrompt.assertComesFromUser()
     attemptLock()
+    var response: GenerateContentResponse
+    var prompt = inputPrompt
+    val tempHistory = LinkedList<Content>()
     try {
-      val response = model.generateContent(*history.toTypedArray(), prompt)
-      history.add(prompt)
-      history.add(response.candidates.first().content)
+      while (true) {
+        response =
+          model.generateContent(*history.toTypedArray(), *tempHistory.toTypedArray(), prompt)
+        val responsePart = response.candidates.first().content.parts.first()
+
+        tempHistory.add(prompt)
+        tempHistory.add(response.candidates.first().content)
+        if (responsePart is FunctionCallPart) {
+          if (model.requestOptions.disableAutoFunction) {
+            break
+          }
+          val output = model.executeFunction(responsePart)
+          prompt = Content("function", listOf(FunctionResponsePart(responsePart.name, output)))
+        } else {
+          break
+        }
```

Comment:
This was a bit confusing to read, it'd probably be easier to read if the break cases were abstracted out to the same level

```suggestion
        if (responsePart !is FunctionCallPart) break
        if (model.requestOptions.disableAutoFunction) break

        val output = model.executeFunction(responsePart)
        prompt = Content("function", listOf(FunctionResponsePart(responsePart.name, output)))
```